### ISSUE TITLE
Python Rules Swallow Error Messages

### ISSUE THREAD

**Issue by aaronspringer-parc** [jruizgit/rules#58]
Hi,

First, thanks so much for this awesome package. I've been playing around with it for a while and am super impressed. 

One of the issues I'm having is that the rule functions in Python will swallow any errors they encounter rather than bubbling them up. Here's an example where the rule tries to call the function "errorOutHere" which is undefined and fails. Rather than printing any error to console or some notification it just fails silently. Is this intended behavior? Is there a log of these somewhere or perhaps a debug mode to see these issues?

```
from durable.lang import *

with ruleset('coach'):
    @when_all(m.completed != True)
    def trigger_reminders(c):
        print('An error message should follow this line')
        # this function is not defined and should crash
        errorOutHere()
        # this line is never reached
        print('This part is never reached and no error is thrown')

    @when_start
    def start(host):
        host.assert_fact('coach', {'id': 1, 'sid': 1, 'completed': False})

run_all()
```

**Comment by jruizgit** [comment]
Hi, thank you for your interest in durable_rules. Very good question. I debated the unhandled exception problem for a while. After testing the framework in a very large deployment, I concluded saving the exception in the context was the most useful course of action. You can write an exception handler as follows: 

```python
from durable.lang import *

with ruleset('coach'):
    @when_all(m.completed != True)
    def trigger_reminders(c):
        print('An error message should follow this line')
        # this function is not defined and should crash
        errorOutHere()
        # this line is never reached
        print('This part is never reached and no error is thrown')

    @when_all(+s.exception)
    def exception_handler(c):
        print('Error message: {0}'.format(c.s.exception))
        c.s.exception = None

    @when_start
    def start(host):
        host.assert_fact('coach', {'id': 1, 'sid': 1, 'completed': False})

run_all()
```

**Comment by aaronspringer-parc** [comment]
Great! Thank you! Would you mind if I flesh out the docs surrounding basic items like this and sid? 

We're likely going to be using this package for a upcoming deployment and I found getting started with the package quite difficult due to the lack of information surrounding basic items. Other devs besides myself may be working with this so it could be helpful to get them up to speed quicker.

Thanks again!

**Comment by jruizgit** [comment]
No problem, please feel free to modify the docs and send me a pull request. The current docs are very basic, there are a good number of little features I haven't had time to properly document.

Thanks!

**Comment by aaronspringer-parc** [comment]
Any tips on how to log errors in this manner when using the statechart? I've tried adding the exception catching rule to both individual states and the chart as a whole.

I've been using the Redis monitor command and then searching for exceptions but it'd be great to have it work similarly to the above.

Example code:
```
import redis
from durable.lang import *

with statechart('coach'):
    # State state since it has no incoming edges
    with state('s1'):
        @to('s2')
        def startup(c):
            print('s1')

    with state('s2'):
        # if user misses 3 activities in a row, turn on reminders for that user
        @to('s2')
        @when_all(m.completed != True)
        def s2(c):
            print('Before error...')
            errorHere()
            print('Not reached and no error printed')

        @when_all(+s.exception)
        def exception_handler(c):
            print('Error message: {0}'.format(c.s.exception))
            c.s.exception = None

    @when_all(+s.exception)
    def exception_handler(c):
        print('Error message: {0}'.format(c.s.exception))
        c.s.exception = None


    @when_start
    def start(host):
        host.post('coach', {'id': 1, 'sid': 1, 'completed': False})

if __name__ == '__main__':
    redis.StrictRedis().flushall()
    run_all()
```

**Comment by jruizgit** [comment]
Hi, durable_rules supports nested states following the UML semantics (https://en.wikipedia.org/wiki/UML_state_machine). You can use nested states to write an exception handler like the example below.

Note1: when testing the scenario, I found a bug. The state change would not be propagated if an exception was thrown in a synchronous action. I have published a new build addressing the issue (0.33.93).

Note2: the statechart state change event is posted before the new context version event. That is, if the exception happened in s1 and a trigger in s2 is satisfied by the state change, such trigger be run before the exception handler trigger.

```python
from durable.lang import *

with statechart('coach'):
    with state('work'):
        with state('s1'):
            @to('s2')
            def startup(c):
                print('s1')

        with state('s2'):
            @to('s2')
            def s2(c):
                print('s2 before')
                errorHere()
                print('s2 after')

        @to('canceled')
        @when_all(+s.exception)
        def exception_handler(c):
            print('Error message: {0}'.format(c.s.exception))
            c.s.exception = None

    state('canceled')

    @when_start
    def start(host):
        host.post('coach', {'id': 1, 'sid': 1, 'completed': False})

run_all()
```
